### ISSUE TITLE
Improvement: diagnose undefined behavior: within a translation unit the same identifier appears with both internal and external linkage

### ISSUE THREAD

**Issue by pmor13** [llvm/llvm-project#54215]
```
static int x;
void f(void)
{
    int x;
    {
        extern int x;
    }
}
```
Expected diagnostics:
```
t0.c:7:20: error: the same identifier 'x' appears with both internal and external linkage
    7 |         extern int x;
```

**Comment by AaronBallman** [comment]
Agreed that we should issue a diagnostic in this case.

C2x 6.2.2p7: If, within a translation unit, the same identifier appears with both internal and external linkage, the behavior is undefined.
C++2b [dcl.stc]p6: All declarations for a given entity shall give its name the same linkage.

(There's a lot of good examples for test cases in C++ in http://eel.is/c++draft/dcl.stc#example-1)

**Comment by pmor13** [comment]
@jakevossen5 @junaire Any interest / time / motivation in fixing this issue?

**Comment by junaire** [comment]
I can have a try... But I'm not super familiar with this code path. Would you like to explain it a little bit to me? Thanks! @AaronBallman

**Comment by AaronBallman** [comment]
> I can have a try... But I'm not super familiar with this code path. Would you like to explain it a little bit to me? Thanks! @AaronBallman

I think the most logical place I'd start looking is in `Sema::MergeVarDecl()` because that's the place where we know we have a new declaration and we saw a previous declaration of the same name that's also in scope.

**Comment by pmor13** [comment]
@junaire I see it this way: within a translation unit:
1. Obtain a list of all identifiers with internal or external linkages: e.g. `x:I, y:E, z:I, x:E, z:I`
2. For the same identifiers: check that linkages match: either external, either internal.
3. If unmatched cases found, then report as "error: the same identifier 'x' appears with both internal and external linkage (the behavior is undefined)".

**Comment by elhewaty** [comment]
Hello, Is there any one working on this?
I am new to LLVM, I Wish I could help with this.

**Comment by EugeneZelenko** [comment]
@elhewaty: Usual signs of work are filled `Assignees` field or Differential patch mentioned in issue.

**Comment by junaire** [comment]
> Hello, Is there any one working on this? 

I'm sorry lately I've been busy with other stuff and don't look into the issue too much, feel free to take over!

> I am new to LLVM, I Wish I could help with this.

Welcome! don't hesitate to ask questions if you have any problems!

**Comment by elhewaty** [comment]
@junaire 
In `Sema::MergeVarDecl` I found the following condition:

```c++
// Variables with external linkage are analyzed in FinalizeDeclaratorGroup.

  // FIXME: The test for external storage here seems wrong? We still
  // need to check for mismatches.
  if (!New->hasExternalStorage() && !New->isFileVarDecl() &&
      // Don't complain about out-of-line definitions of static members.
      !(Old->getLexicalDeclContext()->isRecord() &&
        !New->getLexicalDeclContext()->isRecord())) {
    Diag(New->getLocation(), diag::err_redefinition) << New->getDeclName();
    Diag(OldLocation, PrevDiag);
    return New->setInvalidDecl();
  }
```

Is this a good point to start from?
If I want to store the variables, How about adding a new class member of the type 
`map<Identifier, unordered_set<linkage>>` and check before adding any ID, check its linkage.

**Comment by pmor13** [comment]
@junaire Just in case you've missed notification. See message above from [elhewaty](https://github.com/elhewaty) dated October 6th.

**Comment by elhewaty** [comment]
Hello @AaronBallman, I didn't see your reply sorry for that. I started contributing to LLVM middle end lately by working on `missing optimization/ instcombine` issues.
I will work on this today, I am interested in LLVM/Clang, and I wish I could be a maintainer soon, 
I don't know how, but I will continue working.
Thanks for your help.

**Comment by erichkeane** [comment]
@elhewaty seems to have not responded to the feedback on the pull-request, so I'm marking this as unassigned so that someone else can pick this up (or he can resume his effort!).

**Comment by pmor13** [comment]
@jakevossen5 @junaire Any interest / time / motivation in fixing this issue? See message above from @erichkeane dated Jan 16th.

**Comment by jakevossen5** [comment]
I can't anymore based on my works policy on OSS contribution, apologies :(

**Comment by shivam-dhir** [comment]
Hi, I would love to work on this issue

**Comment by flash1729** [comment]
hey @AaronBallman @pmor13 - I am working on a fix for this.

The fix lives in `Sema::MergeVarDecl` as discussed above. My approach is based on the insight that `LookupResult::isShadowed()` already tracks whether the found declaration was shadowed by some other declaration that lookup ignored. When that's true and `Old` has internal linkage, the `extern` can't inherit internal linkage per [§6.2.2p4](https://www.iso-9899.info/n1570.html#6.2.2) and then it gets external linkage instead which triggers the [§6.2.2p7](https://www.iso-9899.info/n1570.html#6.2.2) UB.

Implemented check:
```cpp
 if (New->isLocalVarDecl() && New->hasExternalStorage() &&   // block-scope extern
    Old->isFileVarDecl() && Old->hasLinkage() &&            // file-scope with linkage
    Previous.isShadowed() &&                                 // shadow detected
    Old->getFormalLinkage() == Linkage::Internal) {          // specifically internal
    Diag(New->getLocation(), diag::err_internal_extern_mismatch)
        << New->getDeclName();
    Diag(OldLocation, PrevDiag);
    return New->setInvalidDecl();
  }
```
This correctly fires on the UB case and stays silent when extern directly sees the static (no shadow → inherits internal linkage → no conflict).

I have not found any false positives on the cases I tested which include these -  https://godbolt.org/z/8rnsTEPMx . With this fix Clangs behavior matches GCC where this particular issue is fixed.
Opening a PR for the same, do let me know if you want me to include some more specific tests regarding this.

**Comment by AaronBallman** [comment]
Reopening due to the revert
